sys/stdvga.c

```diff
@@ -0,0 +1,165 @@
+#include <pci.h>
+#include <vga.h>
+#include <stdc.h>
+#include <malloc.h>
+
+#define VGA_PALETTE_SIZE (256 * 3)
+
+typedef struct stdvga_device {
+  pci_device_t *pci_device;
+  resource_t mem;
+  resource_t io;
+
+  unsigned int width;
+  unsigned int height;
+
+  uint8_t *palette_buffer;
+  uint8_t *fb_buffer; /* This buffer is only needed because we can't pass uio
+                         directly to the bus. */
+
+  vga_device_t vga;
+} stdvga_device_t;
+
+#define STDVGA_FROM_VGA(vga)                                                   \
+  (stdvga_device_t *)container_of(vga, stdvga_device_t, vga)
+
+/* Detailed information about VGA registers is available at
+   http://www.osdever.net/FreeVGA/vga/vga.htm */
+
+#define VGA_PALETTE_ADDR 0x3C8
+#define VGA_PALETTE_DATA 0x3C9
+#define VGA_AR_ADDR 0x3C0
+#define VGA_AR_PAS 0x20 /* Palette Address Source bit */
+
+/* Bochs VBE. Simplifies VGA graphics mode configuration a great deal. Some
+   documentation is available at http://wiki.osdev.org/Bochs_VBE_Extensions */
+#define VBE_DISPI_INDEX_XRES 0x01
+#define VBE_DISPI_INDEX_YRES 0x02
+#define VBE_DISPI_INDEX_BPP 0x03
+#define VBE_DISPI_INDEX_ENABLE 0x04
+#define VBE_DISPI_ENABLED 0x01 /* VBE Enabled bit */
+
+/* Offsets for accessing ioports via PCI BAR1 (MMIO) */
+#define VGA_MMIO_OFFSET (0x400 - 0x3c0)
+#define VBE_MMIO_OFFSET 0x500
+
+/* The general overview of the QEMU std vga device is available at
+   https://github.com/qemu/qemu/blob/master/docs/specs/standard-vga.txt */
+#define VGA_QEMU_STDVGA_VENDOR_ID 0x1234
+#define VGA_QEMU_STDVGA_DEVICE_ID 0x1111
+
+static void stdvga_io_write(stdvga_device_t *vga, uint16_t reg, uint8_t value) {
+  bus_space_write_1(&vga->io, reg + VGA_MMIO_OFFSET, value);
+}
+static uint8_t stdvga_io_read(stdvga_device_t *vga, uint16_t reg)
+  __attribute__((unused));
```

Comment:
There's `__unused` macro you can use. I guess you could use it in function definition as well.